src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -53,33 +53,48 @@ public GeneratorDriver RunFullGeneration(Compilation compilation, out Compilatio
                 return this;
             }
 
-            // PERF: if the input compilation is the same as the last compilation we saw, and we have a final compilation
-            //       we know nothing can have changed and can just short circuit, returning the already processed final compilation 
-            if (compilation == _state.Compilation && _state.FinalCompilation is object)
-            {
-                outputCompilation = _state.FinalCompilation;
-                return this;
-            }
-
             // run the actual generation
             var state = StateWithPendingEditsApplied(_state);
             var stateBuilder = PooledDictionary<ISourceGenerator, GeneratorState>.GetInstance();
+            var receivers = PooledDictionary<ISourceGenerator, ISyntaxReceiver>.GetInstance();
```

Comment:
Could this just be a list of (ISourceGenerator, ISyntaxReceiver)?

Reply:
Not easily, we look-up later on if the generator has an associated syntax receiver. We'd have to iterate to find it or have some other lookup scheme.

---
In reply to: [395348994](https://github.com/dotnet/roslyn/pull/42507#discussion_r395348994) [](ancestors = 395348994)